### PR TITLE
Add manage menu

### LINKED ISSUE #418: Add "管理功能" navigation dropdown menu
Add a dropdown menu "管理功能 <ChevronRight >" in the left navigation bar, right before the horizontal separator after which mentee list starts. Move the following menus into it:

- 管理用户
- 管理会议
- 学生面试
- 导师面试

The dropdown menu should appear only if the current user is a UserManager, GroupManager or MentorshipManager

### PR DESCRIPTION
**Pull request by ada-only-one-ada**
Fixed #418.
Screenshot 2024-10-18 at 00 14 02
Screenshot 2024-10-18 at 00 14 09

### CODE REVIEW

**Comment by weihanwang** [comment]
1. the "管理功能“ menu should look like the User menu at the bottom of the screen, with an admin-like icon on the left and a right-pointing chevron on the right
2. can you open the dropdown menu on the right of 管理功能 rather than below it?
3. can you do the same for the User menu, where the dropdown appears at the right side of the user name rather than above it?

**Comment by weihanwang** [comment]
Additionally, can you also move 学生档案 to right above 管理功能 and below 资源库？

File: src/components/Sidebar.tsx [hunk 1]
```diff
@@ -58,6 +60,35 @@ export interface SidebarItem {
   roles?: Role | Role[],
 }
 
+export interface ManageMenuItem {
```
Comment (weihanwang) [inline src/components/Sidebar.tsx:63]: `DropdownMenuItem`. Also remove `export`. Can remove `export` from `SidebarItem` in the same PR.

File: src/components/Sidebar.tsx [hunk 2]
```diff
@@ -49,6 +50,7 @@ import { AttachmentIcon } from '@chakra-ui/icons';
 import { PiFlagCheckeredFill } from 'react-icons/pi';
 
 export const sidebarContentMarginTop = 10;
+export const manageMenuPaddingLeft = 10;
```
Comment (weihanwang) [inline src/components/Sidebar.tsx:53]: No need for export

File: src/components/Sidebar.tsx [hunk 3]
```diff
@@ -58,6 +60,35 @@ export interface SidebarItem {
   roles?: Role | Role[],
 }
 
+export interface ManageMenuItem {
+  name: string, 
+  path: string,
+  roles?: Role | Role[],
+}
+
+const manageMenuItems: ManageMenuItem[] = [
```
Comment (weihanwang) [inline src/components/Sidebar.tsx:69]: `managerDropdownMenuItems`

File: src/components/Sidebar.tsx [hunk 4]
```diff
@@ -245,6 +263,31 @@ const Sidebar = ({ onClose, ...rest }: SidebarProps) => {
 
 export default Sidebar;
 
+export function ManageMenu({ user } : {
```
Comment (weihanwang) [inline src/components/Sidebar.tsx:266]: No need for `export`. Rename to `ManagerDropdownMenu`. No need to pass in `user`. Call useUserContext within the function.

File: src/components/Sidebar.tsx [hunk 5]
```diff
@@ -245,6 +263,31 @@ const Sidebar = ({ onClose, ...rest }: SidebarProps) => {
...
+  user: User,
+}) {
+  const roles = user.roles;
+  if (!isPermitted(roles, 
+    ["UserManager", "GroupManager", "MentorshipManager"])) {
```
Comment (weihanwang) [inline src/components/Sidebar.tsx:271]: would these two line combined exceed 80 chars?

File: src/components/Sidebar.tsx (hunk 2 above)
Comment (weihanwang) [reply src/components/Sidebar.tsx:53]: `managerDropdownMenuPaddingLeft`

Comment (weihanwang) [reply src/components/Sidebar.tsx:53]: And why do you need padding? also if this menu needs padding, the User menu should be padded in the same way sharing variables

File: src/components/Sidebar.tsx [hunk 6]
```diff
@@ -50,14 +48,61 @@ import { PiFlagCheckeredFill } from 'react-icons/pi';
...
   regex: RegExp,
   roles?: Role | Role[],
 }
 
+interface DropdownMenuItem {
+  name: string, 
+  path?: string,
+  roles?: Role | Role[],
+  icon?: React.ReactNode,
+  action?: () => void;
```
Comment (weihanwang) [inline src/components/Sidebar.tsx:64]: `action: () => void | string;` and remove `path`

File: src/components/Sidebar.tsx [hunk 7]
```diff
@@ -190,61 +222,84 @@ const Sidebar = ({ onClose, ...rest }: SidebarProps) => {
...
+  
+  if (!isPermitted(roles, 
+    ["UserManager", "GroupManager", "MentorshipManager"])) {
+      return <></>;
+  }
+
+  const filteredItems = managerDropdownMenuItems
+    .filter(item => isPermitted(roles, item.roles));
+  return <Menu placement='right-start'> 
+    <DropdownMenuButton title={"管理功能"} avatarName={""} />
```
Comment (weihanwang) [inline src/components/Sidebar.tsx:274]: manager menu item should use an icon and not an avatar to keep its style consistent with other navigation menu items. Pass in a component and not `avatarName` here. The component should be an Avartar for user menu item and an icon for manager menu item.

Also, the look and feel "管理功能“ text is different from neighboring menu items. You can use the same style as the other menu items for both user and manager menu item.

**Comment by weihanwang** [comment]
Also please test on mobile and different screen sizes to make sure responsive versions work well.

File: src/components/Sidebar.tsx [hunk 8]
```diff
@@ -207,87 +222,84 @@ const Sidebar = ({ onClose, ...rest }: SidebarProps) => {
...
+    <MenuList bg={backgroundColor} borderColor={boderColor} zIndex="1000">
+      {filteredItems.map((item, index) =>
+        <>
+          <MenuItem key={item.path} as={NextLink} href={item.path}>{item.name}
+          </MenuItem>
+          {index < filteredItems.length - 1 && <MenuDivider />}
```
Comment (weihanwang) [inline src/components/Sidebar.tsx:280]: remove this line. no divider is needed here.

File: src/components/Sidebar.tsx [hunk 9]
```diff
@@ -207,87 +222,84 @@ const Sidebar = ({ onClose, ...rest }: SidebarProps) => {
...
+
+  return <Menu placement='right-start'>
+    <DropdownMenuButton title={name} avatarName={name} />
+    <MenuList bg={backgroundColor} borderColor={boderColor} zIndex="1000">
+      {userDropdownMenuItems.map((item, index) =>
+        <>
+          <MenuItem key={item.path} as={item.path ? NextLink : undefined} 
+            href={item.path} onClick={item.action}>{item.icon}{item.name}
+          </MenuItem>
+          {index < userDropdownMenuItems.length - 1 && <MenuDivider />}
```
Comment (weihanwang) [inline src/components/Sidebar.tsx:254]: remove this line. we can remove all menu dividers

File: src/components/Sidebar.tsx [hunk 10]
```diff
@@ -190,61 +222,95 @@ const Sidebar = ({ onClose, ...rest }: SidebarProps) => {
...
+  const name = formatUserName(user.name);
+  const backgroundColor = useColorModeValue('white', 'gray.900');
+  const boderColor = useColorModeValue('gray.200', 'gray.700'); 
+
+  return <Menu placement='right-start'>
+    <DropdownMenuButton title={name} 
+      component={<Avatar size={'sm'} bg="brand.a" color="white" name={name} />} />
+    <MenuList bg={backgroundColor} borderColor={boderColor}>
+      {userDropdownMenuItems.map(item =>
+        <>
```
Comment (weihanwang) [inline src/components/Sidebar.tsx:253]: remove this line

File: src/components/Sidebar.tsx [hunk 11]
```diff
@@ -190,61 +222,95 @@ const Sidebar = ({ onClose, ...rest }: SidebarProps) => {
...
+          </MenuItem>
+        </>)}
+    </MenuList>
+  </Menu>;
+}
+
+const DropdownMenuButton = ({ title, component
+} : { 
+  title: string,
+  component: React.ReactNode | string,
```
Comment (weihanwang) [inline src/components/Sidebar.tsx:299]: change this line to `icon: React.ReactNode,`

File: src/components/Sidebar.tsx [hunk 12]
```diff
@@ -190,61 +222,95 @@ const Sidebar = ({ onClose, ...rest }: SidebarProps) => {
...
+            onClick={typeof item.action === 'function' ? item.action : undefined}
+            >
+              {item.icon}{item.name}
+          </MenuItem>
+        </>)}
+    </MenuList>
+  </Menu>;
+}
+
+function ManagerDropdownMenu() {
```
Comment (weihanwang) [inline src/components/Sidebar.tsx:266]: why not merge ManagerDropdownMenu and UserDropdownMenu to one function `DropdownMenu`, which takes title, icon, and a DropdownMenuItem object as input parameters?

File: src/components/Sidebar.tsx [hunk 13]
```diff
@@ -50,14 +48,60 @@ import { PiFlagCheckeredFill } from 'react-icons/pi';
...
+interface DropdownMenuItem {
+  name: string, 
+  action: (() => void) | string,
+  roles?: Role | Role[],
+  icon?: React.ReactNode,
+}
+
+const managerDropdownMenuItems: DropdownMenuItem[] = [
+  {
+    name: '管理用户',
```
Comment (weihanwang) [inline src/components/Sidebar.tsx:68]: can you re-order the items as follows:

- 学生面试
- 导师面试
- 管理会议
- 管理用户

**Comment by weihanwang** [comment]
also, the spacing between 管理功能 and the neighboring menu item is different that the spacing among other menu items:

image

File: src/components/Sidebar.tsx [hunk 14]
```diff
@@ -190,61 +222,95 @@ const Sidebar = ({ onClose, ...rest }: SidebarProps) => {
...
+} : { 
+  title: string,
+  component: React.ReactNode | string,
+}) => {
+  return <MenuButton marginX={4} marginBottom={4} paddingLeft={4} py={2} 
+    color={"gray.500"} fontWeight="bold" transition="all 0.3s" 
+    _focus={{ boxShadow: 'none' }}>
+    <HStack>
+      {component}
+      <Text display={{ base: 'flex', [sidebarBreakpoint]: 'flex' }}>
```
[the remark this answers is not in the document]
Comment (ada-only-one-ada) [reply src/components/Sidebar.tsx:306]: Actually many codes in this file was copied from the old version of top bar version:
https://github.com/yuanjian-org/app/pull/310/files#diff-af4d99865275e8d65e6aa32966045f97c9fb38dfb08523f8bbef08d230d8731dL178

[the remark this answers is not in the document]
Comment (weihanwang) [reply src/components/Sidebar.tsx:306]: regardless of where it comes form, I don't see how this line is useful (any more). So consider removing it.

File: src/components/Sidebar.tsx [hunk 15]
```diff
@@ -190,61 +222,95 @@ const Sidebar = ({ onClose, ...rest }: SidebarProps) => {
...
   );
 };
 
 export default Sidebar;
 
+function UserDropdownMenu() {
+  const [user] = useUserContext();
+  const name = formatUserName(user.name);
+  const backgroundColor = useColorModeValue('white', 'gray.900');
+  const boderColor = useColorModeValue('gray.200', 'gray.700');
```
[the remark this answers is not in the document]
Comment (ada-only-one-ada) [reply src/components/Sidebar.tsx:246]: Tried but failed with this error：
Screenshot 2024-10-18 at 13 28 43
通常是因为你的 React 组件试图在未正确设置或提供相应的 React Context 的情况下，使用了 Context 相关的 Hook。在你的例子中，useColorModeValue 函数可能依赖于一个 Context 来获取当前的颜色模式（白天或夜间模式），但如果这个 Context 没有在组件的父级或更高层级中正确提供，就会导致这个错误

[the remark this answers is not in the document]
Comment (weihanwang) [reply src/components/Sidebar.tsx:246]: Ah right I didn't noticed it calls `use*` functions. Okay, at least extract ('white', 'gray.900') and ('gray.200', 'gray.700') into common variables, e.g. `const bgColorModeValues = ['white', 'gray.900']; ... const bdColor = useColorModeValue(bgColorModeValues[0], bgColorModeValues[1]);`

File: src/components/Sidebar.tsx (hunk 12 above)
Comment (ada-only-one-ada) [reply src/components/Sidebar.tsx:266]: Screenshot 2024-10-18 at 14 57 02
<Avatar /> require name to be type: string | undefined
but user.name is type: string | null.

File: src/components/Sidebar.tsx [hunk 16]
```diff
@@ -33,31 +32,83 @@ import { useRouter } from 'next/router';
...
 import { AttachmentIcon } from '@chakra-ui/icons';
 import { PiFlagCheckeredFill } from 'react-icons/pi';
+import { componentSpacing } from 'theme/metrics';
 
 export const sidebarContentMarginTop = 10;
+const sidebarPaddingTop = 8;
+const sidebarItemPaddingY = 4;
+const sidebarItemPaddingLeft = 8;
+const bgColorModeValues = ['white', 'gray.900'];
+const bdColorModeValues = ['gray.200', 'gray.700'];
```
Comment (weihanwang) [inline src/components/Sidebar.tsx:55]: do not shorten `border`. `bg` is a well known short-hand form but not `bd`

File: src/components/Sidebar.tsx [hunk 17]
```diff
@@ -33,31 +32,83 @@ import { useRouter } from 'next/router';
...
   name: string,
   icon: React.ComponentType,
   path: string,
   regex: RegExp,
   roles?: Role | Role[],
 }
 
+interface DropdownMenuItem {
+  name: string, 
+  action: (() => void) | string,
```
Comment (weihanwang) [inline src/components/Sidebar.tsx:68]: comment on the meaning of a string vs a function

File: src/components/Sidebar.tsx [hunk 18]
```diff
@@ -130,28 +167,35 @@ function mentorships2Items(mentorships: Mentorship[] | undefined): SidebarItem[]
...
 const Sidebar = ({ onClose, ...rest }: SidebarProps) => {
   const [me] = useUserContext();
+  const userName = formatUserName(me.name);
   // Save an API call if the user is not a mentor.
   const { data: mentorships } = isPermitted(me.roles, "Mentor") ?
     trpcNext.mentorships.listMineAsMentor.useQuery() : { data: undefined };
   const mentorshipItems = mentorships2Items(mentorships);
+  const backgroundColor = useColorModeValue(bgColorModeValues[0], 
+    bgColorModeValues[1]);
+  const boderColor = useColorModeValue(bdColorModeValues[0],
```
Comment (weihanwang) [inline src/components/Sidebar.tsx:183]: `borderColor`

File: src/components/Sidebar.tsx [hunk 19]
```diff
@@ -130,28 +167,35 @@ function mentorships2Items(mentorships: Mentorship[] | undefined): SidebarItem[]
...
       transition="3s ease"
-      bg={useColorModeValue('white', 'gray.900')}
+      bg={backgroundColor}
       borderRight="1px"
-      borderRightColor={useColorModeValue('gray.200', 'gray.700')}
+      borderRightColor={boderColor}
       w={{ base: "full", [sidebarBreakpoint]: sidebarWidth }}
       pos="fixed"
       h="full"
+      // set pos to fixed creates a new stacking context,
```
Comment (weihanwang) [inline src/components/Sidebar.tsx:195]: `setting`

File: src/components/Sidebar.tsx [hunk 20]
```diff
@@ -190,61 +235,84 @@ const Sidebar = ({ onClose, ...rest }: SidebarProps) => {
...
+  const borderColor = useColorModeValue(bdColorModeValues[0], 
+    bdColorModeValues[1]);
+  const filteredItems = rolesRequired ? 
+    managerDropdownMenuItems.filter(item => isPermitted(user.roles, item.roles))
+    :
+    menuItems;
+
+  if (filteredItems.length === 0) {
+    return <></>;
+  }
```
Comment (weihanwang) [inline src/components/Sidebar.tsx:283]: Remove `rolesRequired`. Instead, use DropdownMenuItem.roles field to filter each item.

File: src/components/Sidebar.tsx [hunk 21]
```diff
@@ -190,61 +235,84 @@ const Sidebar = ({ onClose, ...rest }: SidebarProps) => {
...
+  return <Flex paddingY={sidebarItemPaddingY}>
+    <Menu placement='right-start'>
+    <DropdownMenuButton title={title} icon={icon}/>
+    <MenuList bg={backgroundColor} borderColor={borderColor}>
+      {filteredItems.map((item, index) =>
+        <MenuItem 
+          key={index}
+          as={typeof item.action === 'string' ? NextLink : undefined}
```
Comment (weihanwang) [inline src/components/Sidebar.tsx:291]: define an variable: `const isUrl = typeof item.action === 'string'` and use this variable in line 291 through 293.

File: src/components/Sidebar.tsx [hunk 22]
```diff
@@ -190,61 +235,84 @@ const Sidebar = ({ onClose, ...rest }: SidebarProps) => {
...
+          onClick={typeof item.action === 'function' ? item.action : undefined} >
+          {item.icon}{item.name}
+        </MenuItem>)}
+    </MenuList>
+  </Menu>
+  </Flex>;
+}
+
+const DropdownMenuButton = ({ title, icon
+} : {
```
Comment (weihanwang) [inline src/components/Sidebar.tsx:302]: can merge this line with the above line

File: src/components/Sidebar.tsx [hunk 23]
```diff
@@ -190,61 +235,84 @@ const Sidebar = ({ onClose, ...rest }: SidebarProps) => {
...
+} : { 
+  title: string,
+  icon: React.ReactNode,
+}) => {
+  return <MenuButton marginX={componentSpacing}  paddingLeft={componentSpacing}
+    color={siderbarTextColor} fontWeight="bold" transition="all 0.3s" 
+    _focus={{ boxShadow: 'none' }}>
+    <HStack>
+      {icon}<Text>{title}</Text>
+      <Box><FiChevronRight /></Box>
```
Comment (weihanwang) [inline src/components/Sidebar.tsx:311]: why do you need a Box here?

**Comment by weihanwang** [comment]
There is a bug: On mobile interface, the navigation panel doesn't retreat after an item in a dropdown menu is clicked.

File: src/components/Sidebar.tsx [hunk 24]
```diff
@@ -262,54 +264,54 @@ const Sidebar = ({ onClose, ...rest }: SidebarProps) => {
...
-          {item.icon}{item.name}
-        </MenuItem>)}
+      {filteredItems.map((item, index) => {
+        const isUrl = typeof item.action === 'string';
+        return (
+          <MenuItem 
+            key={index} 
+            // Only sets the link it is a url 
+            {...(isUrl && { as: NextLink })}
+            {...(isUrl && { href: item.action })}
```
Comment (weihanwang) [inline src/components/Sidebar.tsx:294]: `{...isUrl && { as: NextLink,  href: item.action } }`

File: src/components/Sidebar.tsx [hunk 25]
```diff
@@ -262,54 +264,54 @@ const Sidebar = ({ onClose, ...rest }: SidebarProps) => {
...
+            onClick={() => {
+              if (typeof item.action === 'function') item.action();
```
Comment (weihanwang) [inline src/components/Sidebar.tsx:296]: `if (!usUrl) ...`

Comment (ada-only-one-ada) [reply src/components/Sidebar.tsx:296]: This will not pass the type checking. It may still consider it as as string, we cannot put a () after the string.

Comment (weihanwang) [reply src/components/Sidebar.tsx:296]: can you use `(item as function).action()`?

Comment (ada-only-one-ada) [reply src/components/Sidebar.tsx:296]: ` (item as { action: Function }).action();` works

**Review by weihanwang — APPROVED** [review]

**Pull request merged**